### PR TITLE
CRM-20821 & CRM-20938 - fix bugs in managing premium product images

### PR DESCRIPTION
**Pull request by seancolsen**
### Issues

This PR addresses both 

* [CRM-20821 - Saving an existing premium product breaks the image URLs](https://issues.civicrm.org/jira/browse/CRM-20821)
* [CRM-20938 - The auto-resizing of uploaded premium products images does not preserve aspect ratio](https://issues.civicrm.org/jira/browse/CRM-20938)

Details steps to reproduce are provided in both tickets. 

### Summary

| Before | After |
| -- | -- |
| Editing a premium product (without making changes) breaks its image | Images are preserved while editing |
| Uploading a non-square image for a premium product squishes it into a square, distorting the image | Aspect ratio is preserved while resizing images |
| Long functions, deeply nested code, sparse code comments | Improved code clarity through refactoring |

### Notes

The reason the images were getting broken is because of ffc4d2d06e7e4b2d5c2b31c1291a5bec9af08faf which added logic to "strip protocol and domain from image URLs" but mistakenly ended up being a bit overzealous with its "strip" logic &mdash; it would strip path components off the start of the URL *no matter what*. I improved this logic in 368073f (and added tests to support the new logic).

### CODE REVIEW

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: CRM/Contribute/BAO/ManagePremiums.php [hunk 1]
```diff
@@ -86,42 +86,33 @@ public static function setIsActive($id, $is_active) {
...
+      'is_deductible' => FALSE,
+      'currency' => CRM_Core_Config::singleton()->defaultCurrency,
+    );
+    $params = array_merge($defaults, $params);
 
-    $params['is_active'] = CRM_Utils_Array::value('is_active', $params, FALSE);
-    $params['is_deductible'] = CRM_Utils_Array::value('is_deductible', $params, FALSE);
+    // Use local URLs for images when possible
+    $params['image'] = CRM_Utils_String::simplifyURL($params['image'], TRUE);
+    $params['thumbnail'] = CRM_Utils_String::simplifyURL($params['thumbnail'], TRUE);
```
Comment (monishdeb) [inline CRM/Contribute/BAO/ManagePremiums.php]: @seanmadsen can we further minimize to 
```
$params = array_merge($params, array(
     'id' => CRM_Utils_Array::value('premium', $ids),
     'image' => CRM_Utils_String::simplifyURL(CRM_Utils_Array::value('image', $params, ''), TRUE),
     'thumbnail' => CRM_Utils_String::simplifyURL(CRM_Utils_Array::value('thumbnail', $params, '', TRUE),
     'is_active' => FALSE,
     'is_deductible' => FALSE,
     'currency' => CRM_Core_Config::singleton()->defaultCurrency,
));
```
??

**Comment by monishdeb** [comment]
@colemanw @eileenmcnaughton I have reviewed the PR and updated the PR with some additional fix. Ensured that there is no unintended regression due to refactoring. Can you please merge this PR?

**Comment by eileenmcnaughton** [comment]
@monishdeb I have read your patch and agree that it adds some tidy ups and the lines all look correct. I understand from your comments that you have reviewed all Sean's code & tested it - so I'm happy that this is good to merge.

@seanmadsen would be great if you can do a final test against master after merge

File: CRM/Contribute/Form/ManagePremiums.php [hunk 2]
```diff
@@ -209,16 +209,14 @@ public function buildQuickForm() {
   public static function formRule($params, $files) {
 
     // If choosing to upload an image, then an image must be provided
-    if (
-      isset($params['imageOption']) &&
-      $params['imageOption'] == 'image' &&
-      empty($files['uploadFile']['name'])
+    if (CRM_Utils_Array::value('imageOption', $params['imageOption']) == 'image'
```
Comment (seanmadsen) [inline CRM/Contribute/Form/ManagePremiums.php]: The code: 
`CRM_Utils_Array::value('imageOption', $params['imageOption'])` 
needs to  be changed to
`CRM_Utils_Array::value('imageOption', $params)`

File: CRM/Contribute/Form/ManagePremiums.php [hunk 3]
```diff
@@ -209,16 +209,14 @@ public function buildQuickForm() {
...
+      && empty($files['uploadFile']['name'])
     ) {
       $errors['uploadFile'] = ts('A file must be selected');
     }
 
     // If choosing to use image URLs, then both URLs must be present
-    if (isset($params['imageOption']) && $params['imageOption'] == 'thumbnail') {
+    if (CRM_Utils_Array::value('imageOption', $params['imageOption']) == 'thumbnail') {
```
Comment (seanmadsen) [inline CRM/Contribute/Form/ManagePremiums.php]: The code: 
`CRM_Utils_Array::value('imageOption', $params['imageOption'])` 
needs to  be changed to
`CRM_Utils_Array::value('imageOption', $params)`

File: CRM/Contribute/Form/ManagePremiums.php (hunk 2 above)
Comment (monishdeb) [inline CRM/Contribute/Form/ManagePremiums.php]: Ahh right typo .. sorry about that .. going to create a new PR for that

**Comment by monishdeb** [comment]
@seanmadsen I have created a separate PR https://github.com/civicrm/civicrm-core/pull/10761 . Is there anything else I need to fix?

**Comment by seanmadsen** [comment]
@monishdeb PR #10761 is good in that it addresses my comments above. :white_check_mark: 

In further testing I identified another problem with your changes in bb80d2f. I am proposing a fix in #10762. 

I have now completed my testing. Once #10761 and #10762 are merged then I think we can close the book on this little project.

**Pull request merged**